File: packages/deep-engine/src/runtimePackage/dashboardVideoTypes.ts
export const DASHBOARD_VIDEO_REQUIRED_CAPABILITIES = [
  "video-decoder",
  "frame-texture-update",
  "media-clock",
  "playback-controls",
  "seek",
] as const;
export const DASHBOARD_VIDEO_READY_CAPABILITIES = [] as const;
export const DASHBOARD_VIDEO_AUDIO_BLOCKED_CAPABILITIES = ["audio-output"] as const;

export type DashboardVideoCapabilityV1 = typeof DASHBOARD_VIDEO_REQUIRED_CAPABILITIES[number] | "audio-output";

export interface DashboardVideoMediaV1 {
  readonly id: string;
  readonly revision: number;
  readonly format: "mp4-isobmff";
  readonly mime: "video/mp4";
  readonly byteLength: number;
  readonly sha256: string;
  readonly dataBase64: string;
}

/**
 * Native capability contract. Packaged, muted media supports playback,
 * keyboard/pointer controls and bounded seek. Audio remains fail-closed.
 */
export interface DashboardVideoDiagnosticV1 {
  readonly nodeId: string;
  readonly sourceNodeId: string;
  readonly source: {
    readonly uri: string | null;
    readonly availability: "missing" | "external-unresolved" | "packaged";
    readonly packaged: boolean;
    readonly resourceId: string | null;
  };
  readonly playback: {
    readonly fit: "cover" | "contain" | "fill";
    readonly autoplay: boolean;
    readonly muted: boolean;
    readonly loop: boolean;
  };
  readonly state: {
    readonly status: "blocked" | "ready";
    readonly transport: "unavailable" | "autoplay" | "poster";
    readonly positionSeconds: 0;
    readonly durationSeconds: null;
    readonly reason: "source-missing" | "native-video-runtime-unavailable" | "native-video-runtime-ready" | "native-video-audio-unavailable";
    readonly missingCapabilities: readonly DashboardVideoCapabilityV1[];
  };
}

File: packages/deep-engine-native/src/runtime_package/dashboard_video_types.rs
use super::types::required_nullable;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardVideoDiagnostic {
    pub node_id: String,
    pub source_node_id: String,
    pub source: DashboardVideoSource,
    pub playback: DashboardVideoPlayback,
    pub state: DashboardVideoState,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardVideoSource {
    #[serde(deserialize_with = "required_nullable")]
    pub uri: Option<String>,
    pub availability: String,
    pub packaged: bool,
    #[serde(deserialize_with = "required_nullable")]
    pub resource_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardVideoMedia {
    pub id: String,
    pub revision: u64,
    pub format: String,
    pub mime: String,
    pub byte_length: usize,
    pub sha256: String,
    pub data_base64: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardVideoPlayback {
    pub fit: String,
    pub autoplay: bool,
    pub muted: bool,
    pub r#loop: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct DashboardVideoState {
    pub status: String,
    pub transport: String,
    pub position_seconds: f64,
    #[serde(deserialize_with = "required_nullable")]
    pub duration_seconds: Option<f64>,
    pub reason: String,
    pub missing_capabilities: Vec<String>,
}
